File: pos-backend/src/main/java/com/angelica/pos/inventory/movement/mapper/InventoryMovementMapper.java
package com.angelica.pos.inventory.movement.mapper;

import com.angelica.pos.inventory.movement.dto.InventoryMovementResponse;
import com.angelica.pos.inventory.movement.dto.ManualInventoryMovementRequest;
import com.angelica.pos.inventory.movement.entity.InventoryMovement;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

import java.util.List;

@Mapper(componentModel = "spring")
public interface InventoryMovementMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "product", ignore = true)
    @Mapping(target = "createdBy", ignore = true)
    @Mapping(target = "direction", ignore = true)
    @Mapping(target = "type", ignore = true)
    @Mapping(target = "previousStock", ignore = true)
    @Mapping(target = "newStock", ignore = true)
    @Mapping(target = "sourceType", ignore = true)
    @Mapping(target = "sourceId", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    InventoryMovement toEntity(ManualInventoryMovementRequest request);

    @Mapping(target = "productId", source = "product.id")
    @Mapping(target = "productBarcode", source = "product.barcode")
    @Mapping(target = "productName", source = "product.name")
    @Mapping(target = "productUnit", source = "product.unit")
    @Mapping(target = "createdById", source = "createdBy.id")
    @Mapping(target = "createdByUsername", source = "createdBy.username")
    InventoryMovementResponse toResponse(InventoryMovement inventoryMovement);

    List<InventoryMovementResponse> toResponseList(List<InventoryMovement> inventoryMovements);
}

File: pos-frontend/src/features/inventory/movement/infrastructure/mappers/InventoryMovementMapper.ts
import type { PageResponse } from '../../../../../shared/types/PageResponse'
import type {
  InventoryMovement,
  InventoryMovementDirection,
  InventoryMovementType,
  ManualInventoryMovementData,
} from '../../domain/entities/InventoryMovement'

export type BackendInventoryMovementResponse = {
  id: number
  productId: number
  productBarcode: string
  productName: string
  productUnit: string
  createdById: number
  createdByUsername: string
  direction: InventoryMovementDirection
  type: InventoryMovementType
  quantity: number
  previousStock: number
  newStock: number
  description: string
  sourceType: string | null
  sourceId: number | null
  createdAt: string
}

export type BackendManualInventoryMovementRequest = {
  productId: number
  quantity: number
  description: string
}

export const InventoryMovementMapper = {
  toEntity(response: BackendInventoryMovementResponse): InventoryMovement {
    return {
      id: response.id,
      productId: response.productId,
      productBarcode: response.productBarcode,
      productName: response.productName,
      productUnit: response.productUnit,
      createdById: response.createdById,
      createdByUsername: response.createdByUsername,
      direction: response.direction,
      type: response.type,
      quantity: Number(response.quantity),
      previousStock: Number(response.previousStock),
      newStock: Number(response.newStock),
      description: response.description,
      sourceType: response.sourceType,
      sourceId: response.sourceId,
      createdAt: response.createdAt,
    }
  },

  toPage(response: PageResponse<BackendInventoryMovementResponse>): PageResponse<InventoryMovement> {
    return {
      ...response,
      content: response.content.map(InventoryMovementMapper.toEntity),
    }
  },

  toRequest(data: ManualInventoryMovementData): BackendManualInventoryMovementRequest {
    return {
      productId: data.productId,
      quantity: data.quantity,
      description: data.description.trim(),
    }
  },
}
